src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/MapCode/MapCodeEndpoint.cs

```diff
@@ -0,0 +1,573 @@
+        {
+            return false;
+        }
+
+        var mappingSuccess = await TryMapCodeAsync(
+            locations, nodesToMap, changes, documentContext, cancellationToken).ConfigureAwait(false);
+        if (!mappingSuccess)
+        {
+            return false;
+        }
+
+        MergeEdits(changes);
+        return true;
+    }
+
+    private async Task<bool> TryMapCodeAsync(
+        LSP.Location[][] focusLocations,
+        List<SyntaxNode> nodesToMap,
+        ImmutableArray<TextDocumentEdit>.Builder changes,
+        VersionedDocumentContext documentContext,
+        CancellationToken cancellationToken)
+    {
+        var didCalculateCSharpFocusLocations = false;
+        var csharpFocusLocations = new LSP.Location[focusLocations.Length][];
+
+        // We attempt to map the code using each focus location in order of priority.
+        // The outer array is an ordered priority list (from highest to lowest priority),
+        // and the inner array is a list of locations that have the same priority.
+        // If we can successfully map using the first location, we'll stop and return.
+        var mappingSuccess = false;
+        foreach (var locationByPriority in focusLocations)
+        {
+            foreach (var location in locationByPriority)
+            {
+                // The current assumption is that all focus locations will always be in the same document
+                // as the code to map. The client is currently implemented using this behavior, but if it
+                // ever changes, we'll need to update this code to account for it (i.e., take into account
+                // focus location URIs).
+                Debug.Assert(location.Uri == documentContext.Uri);
+
+                var syntaxTree = await documentContext.GetSyntaxTreeAsync(cancellationToken).ConfigureAwait(false);
+                if (syntaxTree is null)
+                {
+                    continue;
+                }
+
+                var razorNodesToMap = new List<SyntaxNode>();
+                foreach (var nodeToMap in nodesToMap)
+                {
+                    // If node is C#, we send it to their language server to handle and ignore it from our end.
+                    if (nodeToMap.IsCSharpNode(out var csharpBody))
+                    {
+                        if (!didCalculateCSharpFocusLocations)
+                        {
+                            csharpFocusLocations = await GetCSharpFocusLocationsAsync(focusLocations, cancellationToken).ConfigureAwait(false);
+                            didCalculateCSharpFocusLocations = true;
+                        }
+
+                        var csharpMappingSuccessful = await TrySendCSharpDelegatedMappingRequestAsync(
+                            documentContext.Identifier, csharpBody, csharpFocusLocations, changes, cancellationToken).ConfigureAwait(false);
+
+                        // If C# delegation fails, we'll default to the client's fallback mapper.
```

Comment:
After talking to Oscar, I updated this logic so we now fall back to the default built-in mapper if C# fails to map